stdlib/effect.ml

```diff
@@ -93,11 +93,13 @@ module Shallow = struct
     let error _ = failwith "impossible" in
     let effc eff k _last_fiber =
       match eff with
-      | M.Initial_setup__ -> raise (E k)
+      | M.Initial_setup__ -> raise_notrace (E k)
       | _ -> error ()
     in
     let s = alloc_stack error error effc in
-    try Obj.magic (runstack s f' ()) with E k -> k
```

Comment:
> taken together they accelerate more than they slow down

Indeed, but with just the third commit and without the second it presumably speeds it up even more! Do we need to worry about `Obj.magic` in the implementation of the runtime?

Reply:
Yes, we always need to worry about `Obj.magic`, especially thanks to the people who will add it to their code and say "it's alright, the standard library uses it as well".

(Note: the performance impact of the two commits is not comparable: the "second commit" (which removes Obj.magic) has an extra jump or something, this is unlikely to be noticeable. The "third commit" (which uses `raise_notrace`) avoids walking through the stack (in the common backtrace-enabled configuration), this is likely to be noticeable.

Reply:
I was only responding to your observation that they accelerate more than they slow down!

I personally draw a distinction between code in the Standard Library which implements data structures, algorithms and other stuff you might want to reimplement or use as an example elsewhere and then a module like this which is implementing part of the runtime. We could, for example, push more of the implementation back into the primitive itself, at which point the "magic" simply becomes the type declaration in the `external` definition. `Obj.magic` in user code is of course a dreadful idea because it's circumventing the type system, but here it's used because we're on the boundary of the (comparatively) untyped internals of the C part of the runtime and its interface with actual OCaml programs.

Reply:
I don't really agree with the argument for this specific use of `Obj.magic`. I think it's there because the person who wrote the code didn't bother thinking harder about how to make the compiler realize that, yes, the typing is correct here. I would agree with you if it was there for a fundamental reason -- a lack of expressivity of the typed encoding of effect handlers, which at some point I suspected was the reason. But the fact that a simple three-line change makes the code entirely type-safe proves that this is not the situation here.

Reply:
I don't see a performance difference between the two versions.  With `Obj.magic` the compiler generates a branch after the call, branching to the return point of the function.  With `error ()` it generates a tailcall to the `error` function.  I assume that neither branch is ever executed.  And the difference in code size is really small.  So, by all means let's avoid `Obj.magic`.

Reply:
@kayceesrk as the original author of this code I believe, I wonder if you can comment. (I feel confident that the change is correct -- that `runstack` can never return a value here, but this reasoning involves a lot of new different pieces.)

Reply:
Good fix @gasche. I would gladly avoid the use of `Obj.magic` in this case. The performance shouldn't matter for this change as the cost of installing a handler is dominated by the cost of allocating a new fiber in `alloc_stack`.

More generally, I am for not using `Obj.magic` as they hinder reasoning about the code both for humans and optimising compilers.